doc/gsmsendsms.man
.\" -*- eval: (nroff-mode) -*-
.de TQ
.br
.ns
.TP \\$1
..
.\" Like TP, but if specified indent is more than half
.\" the current line-length - indent, use the default indent.
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.lu-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GSMSENDSMS 8 "##DATE##" "gsmsendsms v##VERSION##"
.PP
.SH NAME
gsmsendsms \- SMS message sender utility
.PP
.SH SYNOPSIS
.B gsmsendsms
[ \fB-b\fP \fIbaudrate\fP ]
[ \fB--baudrate\fP \fIbaudrate\fP ]
[ \fB-C\fP \fIservice centre address\fP ]
[ \fB--sca\fP \fIservice centre address\fP ]
[ \fB-d\fP \fIdevice\fP ]
[ \fB--device\fP \fIdevice\fP ]
[ \fB-h\fP ]
[ \fB--help\fP ]
[ \fB-I\fP \fIinit string\fP ]
[ \fB--init\fP \fIinit string\fP ]
[ \fB-r\fP ]
[ \fB--requeststat\fP ]
[ \fB-t\fP ]
[ \fB--test\fP ]
[ \fB-v\fP ]
[ \fB--version\fP ]
[ \fB-X\fP ]
[ \fB--xonxoff\fP ]
\fIphonenumber\fP
[ \fItext\fP ]
.PP
.SH DESCRIPTION
\fIgsmsendsms\fP sends SMS short messages using an GSM mobile phone.
.PP
\fIgsmsendsms\fP attaches itself to the \fIdevice\fP given on the command
line (usually an GSM modem) using the specified \fIbaudrate\fP.  If no
\fIdevice\fP is given, the device \fI/dev/mobilephone\fP is used. If
no \fIbaudrate\fP is given, a default baud rate of 38400 is used.
.PP
\fIgsmsendsms\fP accepts a phone number (recipient address) and the
short message text as parameters. The text may have a maximum length
of 160 characters which is the maximum SMS message length. The GSM
default alphabet is used for encoding. ASCII and Latin-1 characters
that can not be encoded using the GSM default alphabet are converted
to the GSM delta character (GSM code 16).
.PP
Error messages are printed to the standard error output.  If the program
terminates on error the error code 1 is returned.
.PP
.SH OPTIONS
.TP
\fB-b\fP \fIbaudrate\fP, \fB--baudrate\fP \fIbaudrate\fP
The baud rate to use.
.TP
\fB-C\fP \fIservice centre address\fP, \fB--sca\fP \fIservice centre address\fP
Sets the service centre address to use for all SUBMIT SMSs (may not
work with some phones).
.TP
\fB-d\fP \fIdevice\fP, \fB--device\fP \fIdevice\fP
The device to which the GSM modem is connected. The default is
\fI/dev/mobilephone\fP.
.TP
\fB-h\fP, \fB--help\fP
Prints an option summary.
.TP
\fB-I\fP \fIinit string\fP, \fB--init\fP \fIinit string\fP
Initialization string to send to the TA (default: "E0"). Note that the
sequence "ATZ" is sent first.
.TP
\fB-r\fP, \fB--requeststat\fP
Request status reports for sent SMS.
.TP
\fB-t\fP, \fB--test\fP
If this option is given the text is converted
to the GSM default alphabet and back to Latin-1. This option can be
used to find out how ASCII or Latin-1 texts are converted to the GSM
default alphabet. Characters that can not be converted to the GSM default
alphabet are reported as ASCII code 172 (Latin-1 boolean "not")
after this double conversion. No SMS messages are sent, a connection
to a mobile phone is not established.
.TP
\fB-v\fP, \fB--version\fP
Prints the program version.
.TP
\fB-X\fP, \fB--xonxoff\fP
Uses software handshaking (XON/XOFF) for accessing the device.
.PP
.SH EXAMPLES
The following two invocations of \fIgsmsendsms\fP each send the same
SMS message to the number "1234":
.PP
.nf
gsmsendsms -d /dev/ttyS2 -b 19200  1234 "This is a test."
echo "This is a test." | gsmsendsms -d /dev/ttyS2 -b 19200  1234
.fi
.PP
.SH FILES
.TP 1.4i
.B /dev/mobilephone
Default mobile phone device.
.PP
.SH AUTHOR
Peter Hofmann <software@pxh.de>
.PP
.SH BUGS
Report bugs to software@pxh.de.  Include a complete, self-contained
example that will allow the bug to be reproduced, and say which
version of \fIgsmsendsms\fP you are using.
.PP
.SH COPYRIGHT
Copyright \(co 1999 Peter Hofmann
.LP
\fIgsmsendsms\fP is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.
.LP
\fIgsmsendsms\fP is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.
.LP
You should have received a copy of the GNU Library General Public License along
with \fIgsmsendsms\fP; see the file COPYING.  If not, write to the Free Software
Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.SH "SEE ALSO"
.BR gsminfo(7),
.BR gsmpb(1),
.BR gsmctl(1),
.BR gsmsmsd(8),
.BR gsmsmsstore(1).



ChangeLog
gsmlib-1.7
     - fixed bug with calculation of userData length if userDataHeader
       is present (octet count was subtracted, not septet count)

     - fixed problem with string erase() at end of PDU for Falcom A2-1

     - -t/--charset option of gsmpb did not work due to missing parameter
       of getopt_long, fixed

     - The SMS decoder/encoder can now handle alphanumeric addresses
       in the GSM default alphabet

     - set only those SMS stores that are actually needed to perform
       SMS store operation

     - Fixed signalling error 321 (Invalid memory index) when trying to
       read from empty SMS store entry

     - Fixed parsing error when reading current network operator if no
       network connection

     - Added capability to parse cell broadcast messages to gsmlib and
       the gsmsmsd program

     - Added workaround for Motorola Timeport 260 bug that doesn't correctly
       report the message status when retrieving SMS messages from the
       store

     - Added workaround for Motorola Timeport 260 that allocates index
       numbers to SMS messages linearly so that index number can be
       be larger than capacity reported by AT command

gsmlib-1.6
     - more fixes for the COPS=? return format

     - fixed putBack() behaviour in gsm_parser (don't put back if end-of-
       stream is reached)

     - added toString() function to gsm_sms_codec's Address class

     - SMS dates and times are now output in a locale-specific manner

     - fixed bug in Parser::getEol() (_eos was accidentally set to true)

     - added gsm_win32_serial module, Win32 project (VC++),  and
       Option FirstFone changes contributed by Frediano Ziglio
       <freddy77@angelfire.com>

     - fix in COM port recognition for Win32 (gsm_util)

     - renamed library files libgsm.* to libgsmme.* (now starting with
       version 1.0) because of conflict with another Debian package

     - in gsm_phonebook and gsm_store the caching of entries can now
       be disabled

     - added facilities to use other character sets for phonebooks

     - fixed workaround for Ericcson SH888 (missing service centre address)

     - fixed bug in gsm_phonebook that prevented texts with the
       character '@' to be written to to the phonebook

     - fixed nasty memory allocation bug in gsmpb/gsmsmsstore
       (automatic MeTa variable went out of scope even though used
       later)

     - fixed Y2K problem in timestamp printing

     - fixed "make dist" to include win32 files

     - added workarounds for Falcom A2-1 (autodetection and enabling by
       "export GSMLIB_FALCOM_A2_1_FIX=1", zero after PDU)

     - fixed bug that caused gsmlib to abort with an assert if a malformed
       PDU was read

gsmlib-1.5
     - adapted MeTa::getCurrentOPInfo() and MeTa::getAvailableOPInfo()
       to handle Nokia 8290 quirks

     - code to set line speed in gsm_unix_serial reinserted (it was
       accidentally removed in previous version)

     - minor changes to initialization sequence in gsm_unix_serial

     - bugfix in gsm_unix_serial.cc: readByte() == 0 does not mean no
       more bytes, but is legal value

     - additionally allowed characters "*#pwPW" and '+' at any
       position in telephone numbers

     - added environment variable GSMLIB_SH888_FIX that (if set)
       enables the gsmlib workaround for Ericsson SH888's broken SMS TPDUs

     - fixed command line parameter handling bug in gsmsmsstore

     - fixed %files section .spec file to correctly include manual pages

     - fixed some bugs in terminal line setup (gsm_unix_serial) regarding
       software/hardware handshake 

     - added new "--sca" option to SMS-related apps to set the SMS
       service centre address on the command line (useful if default is not
       set correctly in the phone)

     - removed tcflush() call in UnixSerialPort::putLine that broke
       the event system needed for gsmsmsd

gsmlib-1.4
     - more attempts to fix UNIX serial port access

     - allow custom backends for sorted phonebooks to be integrated
       into gsmlib (eg. for RDBMS or LDAP storage). Introduced a new
       module gsm_sorted_phonebook_base that contains the infrastructure for
       this.

     - Now gsmlib needs at least gcc-2.95.2 to compile correctly.

     - Implemented option to open device with software handshaking (XON/XOFF).
       The applications now have an -X option to turn this on.

     - gsmlib now contains a facility to interrupt ongoing activity in a
       controlled way. gsm_unix_serial now blocks for one second
       maximum until it checks whether it was interrupted.

     - various small bugfixes

     - added workaround for phones that omit ':' in AT command responses

     - all debugging output is now printed to stderr

     - gsm_unix_serial: new attempt to initialize modem in a more
       robust way (contributed)

     - gsm_sms: fixed handling of user data header (contributed)

gsmlib-1.3
     - fixed bug that caused gsmlib to hang if TAs don't respond
       to AT inquiries for serial number etc.

     - fixed bug with handling of CB mode AT command construction in
       gsm_me_ta.cc, setSMSRoutingToTA()

     - allow '+' as the first character of phonenumbers

     - implemented reading and writing from/to stdin/stdout in
       gsm_sorted_[sms_store|phonebook] and gsmpb/gsmsmsstore

     - fixed (hopefully) the intermittent hangup problem in
       gsm_unix_serial_port

     - Some mobile phones cannot report the status of some facility
       locks. The gsmctl program now prints out "unknown" in the
       corresponding result line from the FLSTAT
       operation if this problem occurs (instead of terminating).

     - switched off non-blocking access to serial device

gsmlib-1.2
     - gsmlib now also works with TAs that can not switch off echo

     - fixed bug that prevented gsmlib from copying SMS_DELIVER and 
       SMS_STATUS_REPORT back to the ME


apps/gsmsendsms.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmsendsms.cc
// *
// * Purpose: GSM sms send program
// *
// * Author:  Peter Hofmann (software@pxh.de)
// *
// * Created: 16.7.1999
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsmlib/gsm_util.h>
#include <iostream>

using namespace std;
using namespace gsmlib;

// options

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"sca", required_argument, (int*)NULL, 'C'},
  {"device", required_argument, (int*)NULL, 'd'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"test", no_argument, (int*)NULL, 't'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

// convert /r and /n to CR and LF

static string unescapeString(char *line)
{
  string result;
  bool escaped = false;
  int index = 0;

  while (line[index] != 0 &&
         line[index] != CR && line[index] != LF)
  {
    if (escaped)
    {
      escaped = false;
      if (line[index] == 'r')
        result += CR;
      else if (line[index] == 'n')
        result += LF;
      else if (line[index] == '\\')
        result += '\\';
      else
        result += line[index];
    }
    else
      if (line[index] == '\\')
        escaped = true;
      else
        result += line[index];

    ++index;
  }
  return result;
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    string device = "/dev/mobilephone";
    bool test = false;
    string baudrate;
    Ref<GsmAt> at;
    string initString = DEFAULT_INIT_STRING;
    bool swHandshake = false;
    // service centre address (set on command line)
    string serviceCentreAddress;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "C:I:d:b:thvX", longOpts, &dummy))
          != -1)
      switch (opt)
      {
      case 'C':
        serviceCentreAddress = optarg;
        break;
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'd':
        device = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 't':
        test = true;
        break;
      case 'v':
        cerr << argv[0] << stringPrintf(_(": version %s [compiled %s]"),
                                        VERSION, __DATE__) << endl;
        exit(0);
        break;
      case 'h':
        cerr << argv[0] << _(": [-b baudrate][-C sca][-d device][-h]"
                             "[-I init string]\n"
                             "  [-t][-v][-X] phonenumber [text]") << endl
             << endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << endl
             << _("  -C, --sca         SMS service centre address") << endl
             << _("  -d, --device      sets the destination device to connect "
                  "to") << endl
             << _("  -h, --help        prints this message") << endl
             << _("  -I, --init        device AT init sequence") << endl
             << _("  -t, --test        convert text to GSM alphabet and "
                  "vice\n"
                  "                    versa, no SMS message is sent") << endl
             << _("  -v, --version     prints version and exits")
             << endl
             << _("  -X, --xonxoff     switch on software handshake") << endl
             << endl
             << _("  phonenumber       recipient's phone number") << endl
             << _("  text              optional text of the SMS message\n"
                  "                    if omitted: read from stdin")
             << endl << endl;
        exit(0);
        break;
      case '?':
        throw GsmException(_("unknown option"), ParameterError);
        break;
      }

    if (! test)
    {
      // open the port and ME/TA
      Ref<Port> port = new
#ifdef WIN32
             Win32SerialPort
#else
             UnixSerialPort
#endif
        (device,
         baudrate == "" ? DEFAULT_BAUD_RATE :
         baudRateStrToSpeed(baudrate),
         initString, swHandshake);
      // switch message service level to 1
      // this enables acknowledgement PDUs
      MeTa m(port);
      m.setMessageService(1);

      at = new GsmAt(m);
    }

    // check parameters
    if (optind == argc)
      throw GsmException(_("phone number and text missing"), ParameterError);

    if (optind + 2 < argc)
      throw GsmException(_("more than two parameters given"), ParameterError);
      
    // get phone number
    string phoneNumber = argv[optind];

    // get text
    string text;
    if (optind + 1 == argc)
    {                           // read from stdin
      char s[1000];
      cin.get(s, 1000);
      text = unescapeString(s);
      if (text.length() > 160)
        throw GsmException(_("text is larger than 160 characters"),
                           ParameterError);
    }
    else
      text = argv[optind + 1];

    if (test)
      cout << gsmToLatin1(latin1ToGsm(text)) << endl;
    else
    {
      // send SMS
      Ref<SMSMessage> submitSMS = new SMSSubmitMessage(text, phoneNumber);
      // set service centre address in new submit PDU if requested by user
      if (serviceCentreAddress != "")
      {
        Address sca(serviceCentreAddress);
        submitSMS->setServiceCentreAddress(sca);
      }
      submitSMS->setAt(at);
      Ref<SMSMessage> ackPDU;
      submitSMS->send(ackPDU);

      // print acknowledgement if available
      if (! ackPDU.isnull())
      cout << ackPDU->toString();
    }
  }
  catch (GsmException &ge)
  {
    cerr << argv[0] << _("[ERROR]: ") << ge.what() << endl;
    return 1;
  }
  return 0;
}
